### PR TITLE
Add type hints for `tests/unittest.py`.

### PR DESCRIPTION
**Pull request by richvdh**
In particular, add type hints for `get_success` and friends, which are then helpful in a bunch of places.

Needs a few fixes, so review commit-by-commit.

### CODE REVIEW

File: tests/unittest.py [hunk 1]
```diff
@@ -511,30 +534,36 @@ async def run_bg_updates():
...
         self.reactor.pump([by] * 100)
 
-    def get_success(self, d, by=0.0):
-        deferred: Deferred[TV] = ensureDeferred(d)
+    def get_success(
+        self,
+        d: Awaitable[TV],
+        by: float = 0.0,
+    ) -> TV:
+        deferred: Deferred[TV] = ensureDeferred(d)  # type: ignore[arg-type]
```
Comment (clokep) [inline tests/unittest.py:548]: What does mypy want the type of `ensureDeferred` to be?

File: tests/storage/databases/main/test_lock.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ def test_simple_lock(self):
         """
         # First to acquire this lock, so it should complete
         lock = self.get_success(self.store.try_acquire_lock("name", "key"))
-        self.assertIsNotNone(lock)
+        assert lock is not None
```
Comment (clokep) [inline tests/storage/databases/main/test_lock.py:31]: I find it quite annoying that mypy doesn't handle `assertIs(Not)None` the same way...

python/mypy#5088 for cross-refs.

File: tests/unittest.py [hunk 3]
```diff
@@ -642,11 +671,11 @@ def register_appservice_user(
 
     def login(
         self,
-        username,
-        password,
-        device_id=None,
+        username: str,
+        password: str,
+        device_id: Optional[str] = None,
         custom_headers: Optional[Iterable[Tuple[AnyStr, AnyStr]]] = None,
```
Comment (clokep) [inline tests/unittest.py:677]: Should this be the `CustomHeaderType`?

**Review by clokep — COMMENTED** [review]
Seems good, just a couple of questions!

File: tests/unittest.py (hunk 1 above)
Comment (clokep) [reply tests/unittest.py:548]: Seems it was a `Coroutine` or `Deferred`: https://github.com/twisted/twisted/blob/5c24e99e671c4082a1ddc8dbeb869402294bd0dc/src/twisted/internet/defer.py#L1094-L1096

File: tests/unittest.py (hunk 3 above)
Comment (richvdh) [reply tests/unittest.py:677]: yup!

File: tests/unittest.py (hunk 1 above)
Comment (richvdh) [reply tests/unittest.py:548]: yeah. We end up passing in things here that implement `Coroutine["Deferred[_T]", object, "_T"]`, (so they work in practice) but aren't typed that way (they just implement `Awaitable[_T]`). To keep the tests themselves clean, it's easier just to `ignore[arg-type]`.

File: tests/unittest.py (hunk 3 above)
Comment (richvdh) [inline tests/unittest.py:677]:
```suggestion
        custom_headers: Optional[Iterable[CustomHeaderType]] = None,
```

File: tests/unittest.py (hunk 1 above)
Comment (clokep) [reply tests/unittest.py:548]: That makes sense, thanks!

**Pull request merged**